### PR TITLE
Initial PR for Javascript tier 1 SDK

### PR DESCRIPTION
**Pull request by yzheqing**
This PR includes the basic features of Wavefront tier 1 SDK, which includes:
1. transforming point, histogram, trace and spanLog objects to wavefront formatted strings
2. HTTP reporter to send compressed data to Wavefront server via direct ingestion
3. Wavefront histogram (using TDigest) and snapshot support

Unit tests have been added for most functions, but there are still a couple of issues left to be done:
1. implementation of application tag and heartbeat service.
2. more thorough end to end testing
3. tier 1 SDK usage examples

### CODE REVIEW

**Comment by github-gentleman363** [comment]
@yzheqing I think you're still actively working on this. Please let me know when this PR is ready for review.

**Comment by yzheqing** [comment]
@yjw9012 Just made some minor changes and the PR should be ready for review now. Thanks!

**Comment by aliu-vmware** [comment]
One major comment, which I had mentioned earlier: In its current state, this is not an isomorphic library.  It will only work in Node.  You need to do two things:
1) use an isomorphic network library, like isomorphic-fetch, axios, or superagent.
2) use some kind of UMD bundler (eg, webpack or rollup).

UMD bundling setup can be a little tricky to get right, so it's probably best to look for a boilerplate repo you can fork and modify.

**Review by aliu-vmware — CHANGES_REQUESTED** [review]
One major comment, which I had mentioned earlier: In its current state, this is not an isomorphic library. It will only work in Node. You need to do two things:

1. use an isomorphic network library, like isomorphic-fetch, axios, or superagent.
1. use some kind of UMD bundler (eg, webpack or rollup).

UMD bundling setup can be a little tricky to get right, so it's probably best to look for a boilerplate repo you can fork and modify.

By moving to a compiled model, you will also be able to take advantage of some more advanced features of the language, like `import` syntax, optional chaining, and nullish coalescing.

File: .prettierrc
Comment (aliu-vmware) [inline .prettierrc:2]: Generally speaking, this is now obsolete, and we prefer double-quoted strings.  Single-quote was from the historical practice of generating HTML in Javascript, since HTML attributes are double-quote delimited.

File: src/common/applicationTags.js
Comment (aliu-vmware) [inline src/common/applicationTags.js]: there is no real private/public distinction in javascript, since all your properties are exposed anyway.  We generally just use this.application, and omit the getter.  For classes like this that are basically just data, we tend not to write classes at all, preferring object literals + transformation functions.

File: src/common/heartbeatService.js
Comment (aliu-vmware) [inline src/common/heartbeatService.js]: This should probably be `!Array.isArray(components)`.

File: src/common/heartbeatService.js
Comment (aliu-vmware) [inline src/common/heartbeatService.js:37]: generally more idiomatic to use computed key syntax here:
```
new Map({
[APPLICATION_TAG_KEY]: applicationTags.application,
...
})
```

File: src/common/metrics/registry.js
Comment (aliu-vmware) [inline src/common/metrics/registry.js]: Generally use import syntax for these, but barring that, we use destructuring syntax: 
`const {WavefrontSDKCounter } = require("./counter");`

File: src/common/metrics/registry.js
Comment (aliu-vmware) [inline src/common/metrics/registry.js]: prefer template strings over concatenation.

File: src/common/heartbeatService.js
Comment (aliu-vmware) [inline src/common/heartbeatService.js]: generally speaking, Map is reserved for times when you intend to have non-string keys.  Most code ends up using objects as maps, since they're easier to work with.

File: src/common/metrics/registry.js
Comment (yzheqing) [reply src/common/metrics/registry.js]: Will make sure to change in other parts as well.

File: .prettierrc
Comment (yzheqing) [reply .prettierrc:2]: Thanks for the info! I'll fix this at last since it may involve a lot of code change and might mess up the PR for review purpose :)

File: src/common/applicationTags.js
Comment (yzheqing) [reply src/common/applicationTags.js]: The ApplicationTags class is implemented here to keep track of states of the data. It is also consistent with the Java and Python SDKs. I'm wondering if it's better to switch to object literals + transformation functions? Otherwise I can replace the getter with the properties themselves :)

Comment (github-gentleman363) [reply src/common/applicationTags.js]: hmm I guess there is no right answer. I think it depends on who will work on and maintain this code. I'd let authors/maintainers choose a coding style they're comfortable with unless it's really against how Javascript should be written.

File: src/common/applicationTags.js
Comment (github-gentleman363) [inline src/common/applicationTags.js]: I think it's reversed? Shouldn't it be `this.cluster ? this.cluster : constants.NULL_TAG_VAL`?
I'd leave the same comment for L91 as well.
We can add a test case for cluster and shard in `heartbeat.test.js`.

File: src/common/applicationTags.js
Comment (github-gentleman363) [inline src/common/applicationTags.js]: Use `const` instead of `let`

File: src/common/metrics/metrics.js
Comment (github-gentleman363) [inline src/common/metrics/metrics.js]: Why is this super class empty? Is there a reason?

File: src/common/metrics/registry.js
Comment (github-gentleman363) [inline src/common/metrics/registry.js]: 1. If `timeoutSecs` was 0, this `if` condition wouldn't have been met in JS. I'm not sure if that's an expected behavior.
2. `Date.now()` returns a number in milliseconds. `timeoutSecs` looks like a number in seconds, judging by the name. If so, we shouldn't compare them directly.

File: src/common/metrics/registry.js
Comment (github-gentleman363) [inline src/common/metrics/registry.js]: Use `const`

File: src/common/metrics/registry.js
Comment (github-gentleman363) [inline src/common/metrics/registry.js]: Use `const`

File: src/common/metrics/registry.js
Comment (github-gentleman363) [inline src/common/metrics/registry.js]: Use `const`

File: src/common/metrics/registry.js
Comment (github-gentleman363) [inline src/common/metrics/registry.js]: hmm if `this.metrics` already had an object associated with the `name`, the new object we created and passed as the second parameter would be unused. Can we implement this in a different way in order to avoid unnecessary object instantiations?

File: src/common/heartbeatService.js
Comment (github-gentleman363) [inline src/common/heartbeatService.js:32]: What if `components` is null or undefined? Then, this line will assign an array containing null or undefined to `components`.

File: src/common/heartbeatService.js
Comment (github-gentleman363) [inline src/common/heartbeatService.js]: If `applicationTags` was undefined or null, this line would throw a js error.

File: src/common/heartbeatService.js
Comment (github-gentleman363) [inline src/common/heartbeatService.js]: Would be good to clear the timer in `finally` block in case of error.

File: src/common/heartbeatService.js
Comment (github-gentleman363) [inline src/common/heartbeatService.js]: I'm not sure if this operation is safe. You are deleting an element from the same set you're iterating over.

**Review by github-gentleman363 — COMMENTED** [review]
Just did the first round of review.

File: src/common/applicationTags.js
Comment (aliu-vmware) [reply src/common/applicationTags.js]: I'm using the main wavefront codebase as a guideline, here.  We're gradually moving away from data-only classes, and we're also not implementing getters for anything.

File: src/common/applicationTags.js
Comment (aliu-vmware) [reply src/common/applicationTags.js]: Can also be condensed using the new nullish coalescing operator

File: src/common/metrics/metrics.js
Comment (yzheqing) [reply src/common/metrics/metrics.js]: The polymorphism style here is to be consistent with the Java/Python SDK. Again I think javascript has a lot of differences in terms of the design. So maybe we can get rid of the interface here.

File: src/common/utils.js
Comment (github-gentleman363) [inline src/common/utils.js:5]: Did you implement this yourself? If not, it'd be nice to specify the reference link as an inline comment.

File: src/common/heartbeatService.js
Comment (yzheqing) [reply src/common/heartbeatService.js:32]: hmm it really depends on how we want to instantiate the class to use the heartbeat service. I'm wondering if we're allowed to send a heartbeat without any component specified?

File: src/common/heartbeatService.js
Comment (yzheqing) [reply src/common/heartbeatService.js]: I think applicationTags cannot be null since by default we need to specify `application` and `service` tags. But still, it would be a good practice to optional chaining here :)

File: src/common/metrics/registry.js
Comment (yzheqing) [reply src/common/metrics/registry.js]: that's a good point. will pass in the constructor as a function instead.

File: src/common/heartbeatService.js
Comment (yzheqing) [reply src/common/heartbeatService.js]: that's a good point. I found JS Set safe about deletion while iterating because essentially it does `Replacing it with an empty value rather than deleting it ensures it will not mess up with iterators' positions`.
(https://stackoverflow.com/questions/28306756/is-it-safe-to-delete-elements-in-a-set-while-iterating-with-for-of)

File: src/common/utils.js
Comment (yzheqing) [reply src/common/utils.js:5]: yes I implemented this class here. I'll add a script to generate headers (with author and registry) for all the files once this PR gets merged :)

File: src/common/heartbeatService.js
Comment (yzheqing) [reply src/common/heartbeatService.js]: actually the error has already been caught in `sendMetric`. Will delete the try...catch block

File: src/common/applicationTags.js
Comment (yzheqing) [reply src/common/applicationTags.js]: Changed class implementation to object and transformation function

File: README.md
Comment (aliu-vmware) [inline README.md]: would be nice to put this under the "start" script in package.json, so you can `npm start` this.

File: src/common/metrics/gauge.js
Comment (github-gentleman363) [inline src/common/metrics/gauge.js:11]: Is `supplier` guaranteed to be a function? If it was falsey like `null` or `undefined`, calling `this.getValue()` would throw an error. If not guaranteed, we can provide a default value for the parameter like `constructor(supplier = () => {})`.

File: test/heartbeat.test.js
Comment (github-gentleman363) [inline test/heartbeat.test.js]: nit picky but we can import `describe` and `it` in the same line.

File: test/histogram.test.js
Comment (github-gentleman363) [inline test/histogram.test.js]: const?

File: src/direct.js
Comment (github-gentleman363) [inline src/direct.js:7]: Can these be `import`ed?

File: example.js
Comment (github-gentleman363) [inline example.js]: What's the purpose of this `while` loop? If it was to invoke `send(wavefrontClient)` every 2000ms, why not use `setInterval`?

**Review by github-gentleman363 — COMMENTED** [review]
I'm not very familiar with all the setup-related stuff like `.mocharc.yaml`, `babel.config.js`, `rollup.config.js`, etc. so I'd ask @aliu-vmware to review them.
As for the rest, LGTM with some minor comments.

File: src/direct.js
Comment (yzheqing) [reply src/direct.js:7]: yah I'm struggling about the inconsistency as well. I tried to `import` the libraries but there seems to be some error when it compiles to UMD bundles. The libraries are not compiled to ES5 syntax.

File: package.json
Comment (aliu-vmware) [inline package.json]: you don't need to put server or token in here, npm will pass the rest of the command line into node

File: src/direct.js
Comment (aliu-vmware) [inline src/direct.js:6]: should be able to `import "isomorphic-fetch";` here

File: src/direct.js
Comment (aliu-vmware) [reply src/direct.js:7]: My guess is you need another plugin in your rollup config: https://github.com/rollup/plugins/tree/master/packages/commonjs

File: src/direct.js
Comment (aliu-vmware) [inline src/direct.js]: can we make this data-driven?  This code is very repetitive.
```
const histogramMapping = { _histogramsInvalid : "histograms.invalid", ... etc }
Object.entries(histogramMapping).forEach(([key, value]) => { this[key] = this._sdkMetricsRegistry.newCounter(value) })
```

Similar constructs can be applied across the class.  Leave it to you to decide what level of complexity you're comfortable with.

File: src/direct.js
Comment (aliu-vmware) [inline src/direct.js:127]: How is dataFormat generated? is it user-provided? do we sanitize this value?

File: src/entities/histogram/histogramImpl.js
Comment (aliu-vmware) [inline src/entities/histogram/histogramImpl.js]: it looks like you want this to be: const distributions = this._priorMinuteBinsList.map(x => x.toDistribution())

File: src/entities/histogram/histogramImpl.js
Comment (aliu-vmware) [inline src/entities/histogram/histogramImpl.js:331]: This seems like it would be very memory-intensive for high cardinality datasets. Ex: mean from 1..100K, n = 1 for all of them. 10 buckets means 10x memory consumption.

File: src/entities/histogram/histogramImpl.js
Comment (aliu-vmware) [inline src/entities/histogram/histogramImpl.js]: when would currVal ever be false??

File: src/direct.js
Comment (aliu-vmware) [reply src/direct.js:7]: commonjs will probably also require dynamicRequireTargets

File: src/direct.js
Comment (yzheqing) [reply src/direct.js:127]: It's defined in the code when we schedule the timer.

File: src/entities/histogram/histogramImpl.js
Comment (yzheqing) [reply src/entities/histogram/histogramImpl.js:331]: Yes I agree that this is an expensive operation for large dataset. just trying to be consistent here with the other SDKs.

File: src/direct.js
Comment (yzheqing) [reply src/direct.js]: thanks for the suggestion! changed the lines here

File: src/direct.js
Comment (yzheqing) [reply src/direct.js:7]: @aliu-vmware I've tried to compile with `commonjs` plugin but it seems that `isomorphic-fetch` still cannot compile properly, thus resulting in errors when using the bundle from tier 2 SDK. I'm wondering if we can leave the `require` statement as is for now and probably go back to revise later?

File: src/direct.js
Comment (aliu-vmware) [inline src/direct.js]: Avoid JSON.parse unless you started with JSON.  The below is equivalent.
```
{
  [`_${t}Valid`]: `${t}.valid`
}
```

**Review by aliu-vmware — APPROVED** [review]

File: .prettierrc
Comment (yzheqing) [reply .prettierrc:2]: Will resolve this in the next PR.

File: src/direct.js
Comment (yzheqing) [reply src/direct.js:7]: Since this is a non-trivial fix, I created another story to keep track of this later. We'll require the module as is for now.

**Pull request merged**
